drivers/i2c/i2c_ite_enhance.c

```diff
@@ -568,6 +605,311 @@ static int i2c_enhance_transfer(const struct device *dev, struct i2c_msg *msgs,
+	dlm_base = (uint32_t)data->i2c_cq_mode_rx_dlm & 0xffffff;
+	IT8XXX2_I2C_RAMH2A(base) = (dlm_base >> 16) & 0xff;
+	IT8XXX2_I2C_RAMHA(base) = (dlm_base >> 8) & 0xff;
+	IT8XXX2_I2C_RAMLA(base) = dlm_base & 0xff;
+
+	/* Set "Command Address Register" to get commands. */
+	dlm_base = (uint32_t)data->i2c_cq_mode_tx_dlm & 0xffffff;
+	IT8XXX2_I2C_CMD_ADDH2(base) = (dlm_base >> 16) & 0xff;
+	IT8XXX2_I2C_CMD_ADDH(base) = (dlm_base >> 8) & 0xff;
+	IT8XXX2_I2C_CMD_ADDL(base) = dlm_base & 0xff;
+}
+
+static void enhanced_i2c_cq_write(const struct device *dev)
+{
+	struct i2c_enhance_data *data = dev->data;
+	struct i2c_cq_packet *i2c_cq_pckt;
+	uint8_t num_bit_2_0 = (data->cq_msgs[0].len - 1) & I2C_CQ_CMD_L_NUM_BIT_2_0;
+	uint8_t num_bit_10_3 = ((data->cq_msgs[0].len - 1) >> 3) & 0xff;
+
+	i2c_cq_pckt = (struct i2c_cq_packet *)data->i2c_cq_mode_tx_dlm;
+	/* Set commands in RAM. */
+	i2c_cq_pckt->id = data->addr_16bit << 1;
+	i2c_cq_pckt->cmd_l = I2C_CQ_CMD_L_P | I2C_CQ_CMD_L_E | num_bit_2_0;
+	i2c_cq_pckt->cmd_h = num_bit_10_3;
+	for (int i = 0; i < data->cq_msgs[0].len; i++) {
+		i2c_cq_pckt->wdata[i] = data->cq_msgs[0].buf[i];
+	}
+}
+
+static void enhanced_i2c_cq_read(const struct device *dev)
+{
+	struct i2c_enhance_data *data = dev->data;
+	struct i2c_cq_packet *i2c_cq_pckt;
+	uint8_t num_bit_2_0 = (data->cq_msgs[0].len - 1) & I2C_CQ_CMD_L_NUM_BIT_2_0;
+	uint8_t num_bit_10_3 = ((data->cq_msgs[0].len - 1) >> 3) & 0xff;
+
+	i2c_cq_pckt = (struct i2c_cq_packet *)data->i2c_cq_mode_tx_dlm;
+	/* Set commands in RAM. */
+	i2c_cq_pckt->id = data->addr_16bit << 1;
+	i2c_cq_pckt->cmd_l = I2C_CQ_CMD_L_RW | I2C_CQ_CMD_L_P |
+				I2C_CQ_CMD_L_E | num_bit_2_0;
+	i2c_cq_pckt->cmd_h = num_bit_10_3;
+}
+
+static void enhanced_i2c_cq_write_to_read(const struct device *dev)
+{
+	struct i2c_enhance_data *data = dev->data;
+	struct i2c_cq_packet *i2c_cq_pckt;
+	uint8_t num_bit_2_0 = (data->cq_msgs[0].len - 1) & I2C_CQ_CMD_L_NUM_BIT_2_0;
+	uint8_t num_bit_10_3 = ((data->cq_msgs[0].len - 1) >> 3) & 0xff;
+	int i;
+
+	i2c_cq_pckt = (struct i2c_cq_packet *)data->i2c_cq_mode_tx_dlm;
+	/* Set commands in RAM. (command byte for write) */
+	i2c_cq_pckt->id = data->addr_16bit << 1;
+	i2c_cq_pckt->cmd_l = num_bit_2_0;
+	i2c_cq_pckt->cmd_h = num_bit_10_3;
+	for (i = 0; i < data->cq_msgs[0].len; i++) {
+		i2c_cq_pckt->wdata[i] = data->cq_msgs[0].buf[i];
+	}
+
+	/* Set commands in RAM. (command byte for read) */
+	num_bit_2_0 = (data->cq_msgs[1].len - 1) & I2C_CQ_CMD_L_NUM_BIT_2_0;
+	num_bit_10_3 = ((data->cq_msgs[1].len - 1) >> 3) & 0xff;
+	i2c_cq_pckt->wdata[i++] = I2C_CQ_CMD_L_RS | I2C_CQ_CMD_L_RW |
+				I2C_CQ_CMD_L_P | I2C_CQ_CMD_L_E | num_bit_2_0;
+	i2c_cq_pckt->wdata[i] = num_bit_10_3;
+}
+
+static int enhanced_i2c_cmd_queue_trans(const struct device *dev)
+{
+	struct i2c_enhance_data *data = dev->data;
+	const struct i2c_enhance_config *config = dev->config;
+	uint8_t *base = config->base;
+
+	/* ISR of command queue mode. */
+	if (k_is_in_isr()) {
```

Comment:
This block is relevant only to the ISR, so can it be put in a different function? It seems confusing to call this function from either ISR or thread context but take a completely separate branch depending on the context.

Reply:
Done.